### ISSUE TITLE
Test results are different

### ISSUE THREAD

**Issue by Geministudents** [RahulSChand/gpu_poor#3]
Thk for your contribution
When I use this project for fine tuning：https://github.com/hiyouga/LLaMA-Factory
I used the Baichuan-13B model for sft，max_token=800,the actual memory size I use is: 28G（A40）
BUT Use your project test as:44241MB（43G）
What is the problem that causes such a difference?
Looking forward to your reply

MY TRAIN SCRIPT：
CUDA_VISIBLE_DEVICES=0 python ../src/train_bash.py \
    --stage sft \
    --template baichuan \
    --model_name_or_path /container/LLM/Baichuan-13B-Chat \
    --do_train \
    --dataset test_data \
    --dataset_dir ../data \
    --val_size 0.1 \
    --finetuning_type lora \
    --lora_target W_pack \
    --output_dir output_Bc \
    --overwrite_cache \
    --preprocessing_num_workers 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --lr_scheduler_type cosine \
    --cutoff_len 800 \
    --max_new_tokens 1400 \
    --logging_steps 10 \
    --save_steps 6 \
    --eval_steps 6 \
    --max_grad_norm 0.5 \
    --learning_rate 5e-5 \
    --num_train_epochs 3.0 \
    --evaluation_strategy steps \
    --load_best_model_at_end \
    --plot_loss \
    --fp16 \
    --overwrite_output_dir \
    --seed 3407

![image](https://github.com/RahulSChand/gpu_poor/assets/45708697/d97f86e3-9663-4c47-85c7-5586b376e478)

**Comment by RahulSChand** [comment]
@Geministudents This is most likely due to `gradient_checkpointing` being `True` by default in the Llama-factory code (at line 42 of https://github.com/hiyouga/LLaMA-Factory/blob/main/src/llmtuner/tuner/core/utils.py)


<img width="689" alt="image" src="https://github.com/RahulSChand/gpu_poor/assets/16897807/59f4bbe8-f2ca-4d7d-974f-a8dd12439dcb">


The extra memory the site gives is due to "activations" (18730MB), gradient checkpoint means these intermediate activations (result of matmul with q,k,v,O, ffn1, ffn2, ffn3) are not saved but rather re-computed again during backward. This saves memory but increases run time. The site gives you values with assumption there is NO gradient checkpointing (I will add a feature to include this latter)

You can read more about memory savings due to gradient checkpoint here: 
https://aman.ai/primers/ai/grad-accum-checkpoint/

Also you can try setting that flag as `False` and check your memory requirement they will be higher than 28GB

**Comment by Geministudents** [comment]
Thank you  very much for your reply. I will verify this parameter next working day

**Comment by RahulSChand** [comment]
FYI @Geministudents I have added feature for "Gradient Checkpointing" as well

**Comment by Geministudents** [comment]
Thank you for your improvement, but I tested it today. The error of the Baichuan model is not big, but there is a big error when using the Qianwen-14B model.
![c17a7924ef9614877f71c46bbc450ce](https://github.com/RahulSChand/gpu_poor/assets/45708697/40d432a3-24b3-467a-b244-f83933e03e90)
MY TRAIN SCRIPT：
CUDA_VISIBLE_DEVICES=1 python ../src/train_bash.py \
    --stage sft \
    --template chatml \
    --model_name_or_path /container/LLM/Qwen-14B-Chat \
    --do_train \
    --dataset test_data \
    --dataset_dir ../data \
    --val_size 0.1 \
    --finetuning_type lora \
    --lora_target c_attn \
    --output_dir output_Qw \
    --overwrite_cache \
    --preprocessing_num_workers 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --lr_scheduler_type cosine \
    --cutoff_len 800 \
    --max_new_tokens 1400 \
    --logging_steps 10 \
    --save_steps 6 \
    --eval_steps 6 \
    --max_grad_norm 0.5 \
    --learning_rate 5e-5 \
    --num_train_epochs 3.0 \
    --evaluation_strategy steps \
    --load_best_model_at_end \
    --plot_loss \
    --fp16 \
    --overwrite_output_dir \
    --seed 3407

**Comment by RahulSChand** [comment]
@Geministudents This is because the HuggingFace repo of Qwen-14B has a mistake https://huggingface.co/Qwen/Qwen-14B-Chat/tree/main

The config.json (from which my app derives model size) says that intermediate size is 27392, but in actual model weights,  intermediate size is 13696.



![Screenshot from 2023-11-06 14-13-58](https://github.com/RahulSChand/gpu_poor/assets/16897807/92e7d3fc-0288-4022-bc5a-a2068c487698)

This won't doesn't make a difference when using the model but my app needs the config.json to match the model weights to give correct prediction

**Comment by Geministudents** [comment]
Thanks again for your reply, your project helped me a lot.
I will report this issue to the official website of Qianwen.
